config/initializers/controlled_vocabularies.rb

```diff
@@ -1,35 +1,203 @@
+
+    vocab_items = vocab[vocab_name]
+    raise VocabularyInvalidError, 'Vocab must contain key-value mappings' unless vocab_items.is_a?(Hash)
+
+    vocabularies[vocab_name] = {
+      is_i18n: true,
+      data: nil
+    }
 
-  # this loop consumes the controlled_vocabularies
-  config_files.each do |file| # rubocop:disable Style/CombinableLoops
-    config = YAML.safe_load(File.open(file)).deep_symbolize_keys.freeze
-    raise VocabularyInvalidError, 'There should be only one top-level vocabulary name key' unless config.keys.count == 1
+    uri_mappings = vocab_items.invert
 
-    vocab_name = config.keys.first
-    uri_mappings = config[vocab_name].invert
-    namespace[vocab_name] = OpenStruct.new(config[vocab_name]).tap do |vocab|
-      vocab.define_singleton_method(:from_uri) do |uri|
+    # For i18n vocabs, we create a simple object whose methods are the symbols from the file, which
+    # return the corresponding URI values.
+    #
+    # We additionally extend the object with a method for mapping URIs back to symbols, and extend `method_missing`
+    # to provide better error messages.
+
+    vocabularies[vocab_name][:data] = OpenStruct.new(vocab_items).tap do |vocab_responder|
+      vocab_responder.define_singleton_method(:from_uri) do |uri|
         uri_mappings[uri]
       end
-      vocab.define_singleton_method(:method_missing) do |name, *args, &block|
+      vocab_responder.define_singleton_method(:from_value) do |value|
+        vocab_items[value.to_sym]
+      end
+      vocab_responder.define_singleton_method(:method_missing) do |name, *args, &block|
         super(name, *args, &block) || (raise JupiterCore::VocabularyMissingError, "Unknown #{vocab_name} key: #{name}")
       end
     end
   end
-  controlled_vocabularies[File.basename(dir).to_sym] = namespace
+
+  raw_vocabularies = Dir.glob("#{dir}/*.raw.yml")
+
+  raw_vocabularies.each do |file|
+    vocab = YAML.safe_load(File.open(file)).freeze
+    raise VocabularyInvalidError, 'There should be only one top-level vocabulary name key' unless vocab.keys.count == 1
+
+    vocab_name = vocab.keys.first
+    raise VocabularyInvalidError, "Vocab #{vocab_name} already exists!" if vocabularies.key?(vocab_name)
+
+    vocab_items = vocab[vocab_name]
+    raise VocabularyInvalidError, 'Vocab must contain key-value mappings' unless vocab_items.is_a?(Hash)
+
+    vocab_name = vocab_name.to_sym
+    vocabularies[vocab_name] = {
+      is_i18n: false,
+      data: nil
+    }
+
+    uri_mappings = vocab_items.invert
+
+    # For raw vocabs there is no method lookup, only from_uri and from_value
+    vocabularies[vocab_name][:data] = Class.new do
+      # Ruby does something slightly weird with scope here, easier to give these in-class constants
+      VOCAB_ITEMS = vocab_items
+      URI_MAPPINGS = uri_mappings
+
+      def self.from_uri(uri)
+        URI_MAPPINGS[uri]
+      end
+
+      def self.from_value(value)
+        VOCAB_ITEMS[value]
+      end
```

Comment:
Another alternative implementation.

```suggestion
      define_singleton_method :from_uri |uri| do
        uri_mappings[uri]
      end

      define_method :from_value |value| do
        vocab_items[value]
      end
```

Reply:
Smart observation, that would solve the scope issue I'm papering over with the unnecessary constants. That book is really paying off for you!